src/compiler/scala/tools/nsc/typechecker/SyntheticMethods.scala

```diff
@@ -117,11 +117,14 @@ trait SyntheticMethods extends ast.TreeDSL {
       )
     }
 
-    /* Common code for productElement and (currently disabled) productElementName */
+    /* Common code for productElement and productElementName */
     def perElementMethod(name: Name, returnType: Type)(caseFn: Symbol => Tree): Tree =
       createSwitchMethod(name, accessors.indices, returnType)(idx => caseFn(accessors(idx)))
 
-    // def productElementNameMethod = perElementMethod(nme.productElementName, StringTpe)(x => LIT(x.name.toString))
+    def productElementNameMethod = {
+      val constrParamAccessors = clazz.constrParamAccessors
+      createSwitchMethod(nme.productElementName, constrParamAccessors.indices, StringTpe)(idx => LIT(constrParamAccessors(idx).name.dropLocal.toString))
```

Comment:
Probably should be `.name.dropLocal.decodedName` to fix:

```
scala> case class Symbols(:: : Int, || : Int)
defined class Symbols

scala> Symbols(0, 0).productElementNames.toList
res4: List[String] = List($colon$colon, $bar$bar)
```

Reply:
Thanks. Added a test case for this.